<a href="https://colab.research.google.com/github/fralfaro/MAT281/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT281 - Laboratorio N°03





**Objetivo**: Aplicar técnicas avanzadas de manipulación y análisis de datos con pandas sobre un conjunto real de datos de contenido de Netflix, reforzando buenas prácticas y métodos eficientes sin recurrir a `groupby`, `merge`, `pivot`, ni `join`.



**Dataset**:

Trabajaremos con el archivo `netflix_titles.csv`, que contiene información sobre los títulos disponibles en la plataforma Netflix hasta el año 2021.

| Variable       | Clase     | Descripción                                                                 |
|----------------|-----------|------------------------------------------------------------------------------|
| show_id        | caracter  | Identificador único del título en el catálogo de Netflix.                   |
| type           | caracter  | Tipo de contenido: 'Movie' o 'TV Show'.                                     |
| title          | caracter  | Título del contenido.                                                       |
| director       | caracter  | Nombre del director (puede ser nulo).                                       |
| cast           | caracter  | Lista de actores principales (puede ser nulo).                              |
| country        | caracter  | País o países donde se produjo el contenido.                                |
| date_added     | fecha     | Fecha en la que el título fue agregado al catálogo de Netflix.              |
| release_year   | entero    | Año de lanzamiento original del título.                                     |
| rating         | caracter  | Clasificación por edad (por ejemplo: 'PG-13', 'TV-MA').                      |
| duration       | caracter  | Duración del contenido (minutos o número de temporadas para series).        |
| listed_in      | caracter  | Categorías o géneros en los que está clasificado el contenido.              |
| description    | caracter  | Breve sinopsis del contenido.                                               |




In [172]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...


### Parte 1: Limpieza y preparación

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)



In [173]:
#Punto 1
df.shape

(8807, 12)

El dataset tiene 8807 filas y 12 columnas.

In [174]:
#df.dtypes
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [175]:
print(f"Clase de 'show_id': {type(df["show_id"][0])}")
print(f"Clase de 'type': {type(df["type"][0])}")
print(f"Clase de 'title': {type(df["title"][0])}")
print(f"Clase de 'director': {type(df["director"][0])}")
print(f"Clase de 'cast': {type(df["cast"][1])}")
print(f"Clase de 'duration': {type(df["duration"][0])}")
print(f"Clase de 'listed_in': {type(df["listed_in"][0])}")

Clase de 'show_id': <class 'str'>
Clase de 'type': <class 'str'>
Clase de 'title': <class 'str'>
Clase de 'director': <class 'str'>
Clase de 'cast': <class 'str'>
Clase de 'duration': <class 'str'>
Clase de 'listed_in': <class 'str'>


El dataframe contiene datos de tipo entero (int) y *string* (str).

In [176]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

La columna "director" tiene 2634 datos nulos, siendo la columna con más cantidad de estos datos. Le siguen la columna "country" con 831 datos nulos, la columna "cast" con 825 datos nulos, "date_added" con 10 datos nulos, "rating" con 4 y "duration" con 3. Las demás columnas no tienen valores nulos.

In [177]:
#Punto 2
import datetime
df["date_added"] = pd.to_datetime(df["date_added"], errors="coerce")
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   
4 2021-09-24          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [178]:
#Punto 3
df = df.assign(year_added=df["date_added"].dt.year, month_added=df["date_added"].dt.month)
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   
4 2021-09-24          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  year_added  month_added  
0  As her father nears the end of his life, filmm...      2021.0          9.0  
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0  
2  To protect his family from a powerful drug lor...      2021.0          9.0  
3  Feuds, flirtations and toilet talk go down amo...      2021.0          9.0  
4  In a city of coaching centers known to train I...      2021.0          9.0

## Parte 2: Técnicas avanzadas de pandas

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).
   * Extraer la duración en minutos para las películas desde la columna `duration`.

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

**Suposición:** De ahora en adelante, asumiremos que los puntos son independientes entre sí.

In [179]:
#Punto 4
df_movies = df.loc[(df["type"]=="Movie") & (df["year_added"]>2018)]
df_movies.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country date_added  \
0                                       United States 2021-09-25   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  year_added  month_added  
0   As her father nears the end of his life, filmm...      2021.0          9.0  
6   Equestria's divided. But a bright-eyed hero be...      2021.0          9.0  
7   On a photo shoot in Ghana, an American model s...      2021.0          9.0  
9   A woman adjusting to life after a loss contend...      2021.0          9.0  
12  After most of her family is murdered in a terr...      2021.0          9.0

In [180]:
#Punto 5
df_love = df.loc[df["title"].apply(lambda x: x.lower()).str.contains("love")]
df_love.head()

show_id     type                      title               director  \
25      s26  TV Show       Love on the Spectrum                    NaN   
158    s159    Movie    Love Don't Cost a Thing              Troy Byer   
159    s160    Movie             Love in a Puff         Pang Ho-cheung   
206    s207    Movie  LSD: Love, Sex Aur Dhokha       Dibakar Banerjee   
227    s228    Movie                Really Love  Angel Kristi Williams   

                                                  cast        country  \
25                                    Brooke Satchwell      Australia   
158  Nick Cannon, Christina Milian, Kenan Thompson,...  United States   
159  Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...      Hong Kong   
206  Nushrat Bharucha, Anshuman Jha, Neha Chauhan, ...          India   
227  Kofi Siriboe, Yootha Wong-Loi-Sing, Michael Ea...  United States   

    date_added  release_year rating   duration  \
25  2021-09-21          2021  TV-14  2 Seasons   
158 2021-09-01          2003  PG-13    101 min   
159 2021-09-01          2010  TV-MA    103 min   
206 2021-08-27          2010  TV-MA    112 min   
227 2021-08-25          2020  TV-MA     95 min   

                                            listed_in  \
25     Docuseries, International TV Shows, Reality TV   
158                         Comedies, Romantic Movies   
159            Comedies, Dramas, International Movies   
206  Dramas, Independent Movies, International Movies   
227       Dramas, Independent Movies, Romantic Movies   

                                           description  year_added  \
25   Finding love can be hard for anyone. For young...      2021.0   
158  A nerdy teen tries to make himself cool by ass...      2021.0   
159  When the Hong Kong government enacts a ban on ...      2021.0   
206  This provocative drama examines how the voyeur...      2021.0   
227  A rising Black painter tries to break into a c...      2021.0   

     month_added  
25           9.0  
158          9.0  
159          9.0  
206          8.0  
227          8.0

In [181]:
df_duration=df.assign(duration_2=df["duration"].str.extract(r'(\d+)\s+min').astype(float))
#Este comando retorna NaN para las series, pues su duración está en temporadas en lugar de minutos
df_duration = df_duration.loc[df_duration["type"]=="Movie"]
df_duration.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country date_added  \
0                                       United States 2021-09-25   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  year_added  \
0   As her father nears the end of his life, filmm...      2021.0   
6   Equestria's divided. But a bright-eyed hero be...      2021.0   
7   On a photo shoot in Ghana, an American model s...      2021.0   
9   A woman adjusting to life after a loss contend...      2021.0   
12  After most of her family is murdered in a terr...      2021.0   

    month_added  duration_2  
0           9.0        90.0  
6           9.0        91.0  
7           9.0       125.0  
9           9.0       104.0  
12          9.0       127.0

In [182]:
# Punto 6
#Transformamos datos de columna "listed_in" a listas
df_exp = df.assign(listed_in=df["listed_in"].apply(lambda x: x.split(", ")))
#Explotamos
df_exp = df_exp.explode("listed_in")
df_exp.head()

show_id     type                 title         director  \
0      s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show         Blood & Water              NaN   
1      s2  TV Show         Blood & Water              NaN   
1      s2  TV Show         Blood & Water              NaN   
2      s3  TV Show             Ganglands  Julien Leclercq   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   

  date_added  release_year rating   duration               listed_in  \
0 2021-09-25          2020  PG-13     90 min           Documentaries   
1 2021-09-24          2021  TV-MA  2 Seasons  International TV Shows   
1 2021-09-24          2021  TV-MA  2 Seasons               TV Dramas   
1 2021-09-24          2021  TV-MA  2 Seasons            TV Mysteries   
2 2021-09-24          2021  TV-MA   1 Season          Crime TV Shows   

                                         description  year_added  month_added  
0  As her father nears the end of his life, filmm...      2021.0          9.0  
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0  
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0  
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0  
2  To protect his family from a powerful drug lor...      2021.0          9.0

In [183]:
#Punto 7
#Usamos df_exp del punto anterior para que el conteo de géneros funcione
df_exp["listed_in"].value_counts().head(10) #Imprime el top 10

listed_in
International Movies        2752
Dramas                      2427
Comedies                    1674
International TV Shows      1351
Documentaries                869
Action & Adventure           859
TV Dramas                    763
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Name: count, dtype: int64

In [184]:
#Punto 8
#Reutilizamos la columna "duration_2"
df_long = df.assign(duration_2=df["duration"].str.extract(r'(\d+)\s+min').astype(float))
#Marcamos las películas de más de 120 mins
df_long["long_content"] = df_long["duration_2"].mask(df_long["duration_2"]>120, True)
#Si no son de más de 120 mins, las marcamos como False
df_long["long_content"] = df_long["long_content"].apply(lambda x: False if x!=True else True)
#Borramos la columna auxiliar
df_long = df_long.drop("duration_2", axis=1)
df_long.loc[df_long["type"]=="Movie"].head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country date_added  \
0                                       United States 2021-09-25   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  year_added  \
0   As her father nears the end of his life, filmm...      2021.0   
6   Equestria's divided. But a bright-eyed hero be...      2021.0   
7   On a photo shoot in Ghana, an American model s...      2021.0   
9   A woman adjusting to life after a loss contend...      2021.0   
12  After most of her family is murdered in a terr...      2021.0   

    month_added  long_content  
0           9.0         False  
6           9.0         False  
7           9.0          True  
9           9.0         False  
12          9.0          True

**Suposición:** Para el Punto 9, asumiremos que se pide filtrar las películas que cumplan con las tres condiciones pedidas al mismo tiempo.

In [185]:
#Punto 9
#Reutilizamos la columna "duration_2"
df_filter = df.assign(duration_2=df["duration"].str.extract(r'(\d+)\s+min').astype(float))
df_filter = df_filter.loc[(df_filter["type"]=="Movie") & (df_filter["duration_2"]>100) & (df_filter["rating"]=="R") & (df_filter["country"]=="United States")]
df_filter.head()

show_id   type                        title               director  \
48      s49  Movie                 Training Day          Antoine Fuqua   
81      s82  Movie                         Kate  Cedric Nicolas-Troyan   
131    s132  Movie  Blade Runner: The Final Cut           Ridley Scott   
139    s140  Movie           Do the Right Thing              Spike Lee   
144    s145  Movie                  House Party        Reginald Hudlin   

                                                  cast        country  \
48   Denzel Washington, Ethan Hawke, Scott Glenn, T...  United States   
81   Mary Elizabeth Winstead, Jun Kunimura, Woody H...  United States   
131  Harrison Ford, Rutger Hauer, Sean Young, Edwar...  United States   
139  Danny Aiello, Ossie Davis, Ruby Dee, Richard E...  United States   
144  Christopher Reid, Christopher Martin, Robin Ha...  United States   

    date_added  release_year rating duration  \
48  2021-09-16          2001      R  122 min   
81  2021-09-10          2021      R  106 min   
131 2021-09-01          1982      R  117 min   
139 2021-09-01          1989      R  120 min   
144 2021-09-01          1990      R  104 min   

                                           listed_in  \
48                                 Dramas, Thrillers   
81                                Action & Adventure   
131  Action & Adventure, Classic Movies, Cult Movies   
139                 Classic Movies, Comedies, Dramas   
144                            Comedies, Cult Movies   

                                           description  year_added  \
48   A rookie cop with one day to prove himself to ...      2021.0   
81   Slipped a fatal poison on her final job, a rut...      2021.0   
131  In a smog-choked dystopian Los Angeles, blade ...      2021.0   
139  On a sweltering day in Brooklyn, simmering rac...      2021.0   
144  Grounded by his strict father, Kid risks life ...      2021.0   

     month_added  duration_2  
48           9.0       122.0  
81           9.0       106.0  
131          9.0       117.0  
139          9.0       120.0  
144          9.0       104.0

In [193]:
#Punto 10
#Función para aplicar .style
def make_pretty(styler):
    styler.set_caption("Top 10 Most Long Movies") #Título
    styler.hide() #Quita los índices
    styler.hide(subset=["show_id","type","date_added","year_added","month_added","country","release_year","rating","description","cast","director"], axis=1) #Quita columnas
    headers = {
    'selector': 'th:not(.index_name)',
    'props': 'background-color: #000066; color: white;'
    }
    styler.set_table_styles([headers]) #Colorea la primera fila
    styler.background_gradient(subset=["duration"], cmap='YlGnBu') #Colorea la columna "duration" de acuerdo a sus valores
    return styler

#Reutilizamos la columna "duration_2"
df_top = df.assign(duration=df["duration"].str.extract(r'(\d+)\s+min').astype(float))
#Filtramos por películas
df_top = df_top.loc[df_top["type"]=="Movie"]
#Ordenamos por duración
df_top = df_top.sort_values("duration", ascending=False)
df_top = df_top.head(10)
#aplicamos .style
df_top = df_top.style.pipe(make_pretty)
df_top



### Pregunta Desafío

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).



### Bonus: Análisis de duplicados y limpieza

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?
13. ¿Cuántos títulos únicos hay en total en la columna `title`?





In [187]:
#Punto 11
#Reutilizamos código del punto 6
df_def = df.assign(listed_in=df["listed_in"].apply(lambda x: x.split(", ")))
df_def = df_def.explode("listed_in")
df_def.value_counts(subset=["listed_in", "rating"]).head(10)

listed_in               rating
International Movies    TV-MA     1130
                        TV-14     1065
Dramas                  TV-MA      830
International TV Shows  TV-MA      714
Dramas                  TV-14      693
International TV Shows  TV-14      472
Comedies                TV-14      465
TV Dramas               TV-MA      434
Comedies                TV-MA      431
Dramas                  R          375
Name: count, dtype: int64

Las combinaciones más frecuentes de género y rating son International Movies con TV-MA, International Movies con TV-14 y Dramas con TV-MA.

In [188]:
#Punto 12
#Función para contar cuántas veces aparece un título en la columna "title"
def contar_repetidos(serie,pelicula):
  contador = 0
  for peli in serie:
    if peli.lower() == pelicula.lower():
      contador += 1
  return contador

df_bonus = df.loc[df["type"]=="Movie"]

df_bonus = df_bonus.assign(title_counter = df_bonus["title"].apply(lambda x: contar_repetidos(df_bonus["title"], x)))
df_bonus.loc[df_bonus["title_counter"]>1].head()

show_id   type                 title         director  \
159     s160  Movie        Love in a Puff   Pang Ho-cheung   
303     s304  Movie  Esperando la carroza  Alejandro Doria   
6705   s6706  Movie  Esperando La Carroza  Alejandro Doria   
7345   s7346  Movie        Love In A Puff   Pang Ho-cheung   

                                                   cast    country date_added  \
159   Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...  Hong Kong 2021-09-01   
303   Luis Brandoni, China Zorrilla, Antonio Gasalla...  Argentina 2021-08-05   
6705  Luis Brandoni, China Zorrilla, Antonio Gasalla...  Argentina 2018-07-15   
7345  Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...  Hong Kong 2018-08-01   

      release_year rating duration  \
159           2010  TV-MA  103 min   
303           1985  TV-MA   95 min   
6705          1985     NR   95 min   
7345          2010  TV-MA  103 min   

                                        listed_in  \
159        Comedies, Dramas, International Movies   
303   Comedies, Cult Movies, International Movies   
6705  Comedies, Cult Movies, International Movies   
7345       Comedies, Dramas, International Movies   

                                            description  year_added  \
159   When the Hong Kong government enacts a ban on ...      2021.0   
303   Cora has three sons and a daughter and she´s a...      2021.0   
6705  Cora has three sons and a daughter and she´s a...      2018.0   
7345  When the Hong Kong government enacts a ban on ...      2018.0   

      month_added  title_counter  
159           9.0              2  
303           8.0              2  
6705          7.0              2  
7345          8.0              2

Podemos observar que sí existen películas repetidas con el mismo título. Sin embargo, estas películas tienen el mismo año de lanzamiento. Por tanto, no existen películas con el mismo nombre pero con distinto año de lanzamiento.

In [189]:
#Punto 13
df_bonus2 = df.assign(title = df["title"].apply(lambda x: x.lower()))

df_bonus3 = df.loc[df["type"]=="Movie"]
df_bonus3 = df_bonus3.assign(title = df_bonus3["title"].apply(lambda x: x.lower()))


print(f"Número total de títulos: {len(df_bonus2["title"])}")
print(f"Número de títulos únicas: {len(df_bonus2["title"].unique())}\n")
print(f"Número total de películas: {len(df_bonus3["title"])}")
print(f"Número de películas únicas: {len(df_bonus3["title"].unique())}")

Número total de títulos: 8807
Número de títulos únicas: 8802

Número total de películas: 6131
Número de películas únicas: 6129


Hay 8802 títulos únicos. Asimismo, hay 6129 títulos de películas únicos.